controlplane/kubeadm/internal/controllers/remediation.go

```diff
@@ -191,9 +211,107 @@ func (r *KubeadmControlPlaneReconciler) reconcileUnhealthyMachines(ctx context.C
+// - KCP already reached the maximum number of retries for a machine.
+// NOTE: Counting the number of retries is required In order to prevent infinite remediation e.g. in case the
+// first Control Plane machine is failing due to quota issue.
+func (r *KubeadmControlPlaneReconciler) checkRetryLimits(log logr.Logger, machineToBeRemediated *clusterv1.Machine, controlPlane *internal.ControlPlane) (hasRetries bool, retryCount int32) {
+	// If there is no last remediation, this is the first try of a new retry sequence.
+	if controlPlane.KCP.Status.LastRemediation == nil {
+		return true, 0
+	}
+
+	// Gets MinHealthySeconds and RetryDelaySeconds from the remediation strategy, or use defaults.
+	minHealthyPeriod := controlplanev1.DefaultMinHealthyPeriod
+	if controlPlane.KCP.Spec.RemediationStrategy != nil && controlPlane.KCP.Spec.RemediationStrategy.MinHealthyPeriod != nil {
+		minHealthyPeriod = controlPlane.KCP.Spec.RemediationStrategy.MinHealthyPeriod.Duration
+	}
+
+	retryDelay := time.Duration(0)
+	if controlPlane.KCP.Spec.RemediationStrategy != nil {
+		retryDelay = controlPlane.KCP.Spec.RemediationStrategy.RetryPeriod.Duration
+	}
+
+	// Gets the timestamp of the last remediation; if missing, default to a value
+	// that ensures both MinHealthySeconds and RetryDelaySeconds are expired.
+	// NOTE: this could potentially lead to executing more retries than expected or to executing retries before than
+	// expected, but this is considered acceptable.
+	// when the system recovers from someone/something manually removing the LastRemediatingTimeStampAnnotation.
+	max := func(x, y time.Duration) time.Duration {
+		if x < y {
+			return y
+		}
+		return x
+	}
+
+	lastRemediationTimestamp := time.Now().Add(-2 * max(minHealthyPeriod, retryDelay)).UTC()
+	if controlPlane.KCP.Status.LastRemediation != nil {
+		lastRemediationTimestamp = time.Unix(controlPlane.KCP.Status.LastRemediation.Timestamp.Seconds, int64(controlPlane.KCP.Status.LastRemediation.Timestamp.Nanos))
+	}
+
+	// Check if the machine being remediated has been created as a remediation for a previous unhealthy machine.
+	// NOTE: if someone/something manually removing the RemediationForAnnotation on Machines or one of the LastRemediatedMachineAnnotation
+	// and LastRemediatedMachineRetryAnnotation on KCP, this could potentially lead to executing more retries than expected,
+	// but this is considered acceptable in such a case.
+	machineRemediatingFor := machineToBeRemediated.Name
+	if remediationFor, ok := machineToBeRemediated.Annotations[controlplanev1.RemediationForAnnotation]; ok {
+		// If the remediation is happening before minHealthyPeriod is expired, then KCP considers this
+		// as a remediation for the same previously unhealthy machine.
+		// TODO: add example
+		if lastRemediationTimestamp.Add(minHealthyPeriod).After(time.Now()) {
+			machineRemediatingFor = remediationFor
+			log.WithValues("RemediationRetryFor", klog.KRef(machineToBeRemediated.Namespace, machineRemediatingFor))
```

Comment:
```suggestion
			log = log.WithValues("RemediationRetryFor", klog.KRef(machineToBeRemediated.Namespace, machineRemediatingFor))
```
Same in other places